worlds/jakanddaxter/Rules.py

```diff
@@ -1,102 +1,116 @@
-        return can_reach_orbs_global(state, player, multiworld)
+    if options.enable_orbsanity == EnableOrbsanity.option_off:
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_vanilla(state, player, required_orbs, required_previous_trade))
     else:
-        return can_reach_orbs_level(state, player, multiworld, level_name)
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_orbsanity(state, player, required_orbs, required_previous_trade))
 
 
-def can_reach_orbs_global(state: CollectionState,
-                          player: int,
-                          multiworld: MultiWorld) -> int:
+def recalculate_reachable_orbs(state: CollectionState, player: int) -> None:
+
+    if not state.prog_items[player]["Reachable Orbs Fresh"]:
+
+        # Recalculate every level, every time the cache is stale, because you don't know
+        # when a specific bundle of orbs in one level may unlock access to another.
+        levels = (level for level in Orbs.level_info if level != "")
+        for level in levels:
+            state.prog_items[player][f"{level} Reachable Orbs".strip()] = (
+                count_reachable_orbs_level(state, player, state.multiworld, level))
+
+        # Also recalculate the global count, still used even when Orbsanity is Off.
+        state.prog_items[player]["Reachable Orbs"] = count_reachable_orbs_global(state, player, state.multiworld)
+        state.prog_items[player]["Reachable Orbs Fresh"] = True
+
+
+def count_reachable_orbs_global(state: CollectionState,
+                                player: int,
+                                multiworld: MultiWorld) -> int:
 
     accessible_orbs = 0
     for region in multiworld.get_regions(player):
-        if state.can_reach(region, "Region", player):
-            accessible_orbs += typing.cast(JakAndDaxterRegion, region).orb_count
-
+        if region.can_reach(state):
+            accessible_orbs += typing.cast(JakAndDaxterRegion, region).orb_count  # Only cast when we need to.
     return accessible_orbs
 
 
-def can_reach_orbs_level(state: CollectionState,
-                         player: int,
-                         multiworld: MultiWorld,
-                         level_name: str) -> int:
+def count_reachable_orbs_level(state: CollectionState,
+                               player: int,
+                               multiworld: MultiWorld,
+                               level_name: str = "") -> int:
 
     accessible_orbs = 0
-    regions = [typing.cast(JakAndDaxterRegion, reg) for reg in multiworld.get_regions(player)]
+    regions = typing.cast(typing.List[JakAndDaxterRegion], multiworld.get_regions(player))  # Need to cast all upfront.
     for region in regions:
-        if region.level_name == level_name and state.can_reach(region, "Region", player):
+        if region.level_name == level_name and region.can_reach(state):
```

Comment:
You could test if flipping this conditional around (can reach and level name check) changes gen time from short circuiting better

Reply:
Flipping the condition did not seem to change the gen time, still around 30-40 seconds for Per-Level-1-Orb.